File: node/src/resources/collections.ts
import type { PaydClient } from "../client";
import type {
  MpesaCollectionParams,
  MpesaCollectionResponse,
  CardCollectionParams,
  CardCollectionResponse,
  PanAfricanCollectionParams,
  PanAfricanCollectionResponse,
  BankAccount,
} from "../types/collections";
import { PaydAPIError } from "../errors";
import { validateKenyaPhone, normalizeKenyaPhone } from "../validators/phone";
import { validateInternationalPhone } from "../validators/phone";
import { validateMpesaAmount, validateCardAmount, validatePositiveAmount } from "../validators/amount";
import { validateRequired, validateEnum } from "../validators/params";

/**
 * Collection (payin) operations — collect payments from customers.
 */
export class Collections {
  constructor(private readonly client: PaydClient) {}

  /**
   * Collect a payment via Kenya M-Pesa STK Push.
   *
   * Initiates an STK push to the customer's phone. The customer confirms
   * payment on their device. Final status is delivered via webhook.
   *
   * @example
   * ```typescript
   * const result = await payd.collections.mpesa({
   *   username: "my_payd_user",
   *   amount: 500,
   *   phoneNumber: "0700000000",
   *   narration: "Order #1234",
   *   callbackUrl: "https://example.com/webhook",
   * });
   * console.log(result.transactionReference); // "9BD103350408eR"
   * ```
   */
  async mpesa(params: MpesaCollectionParams): Promise<MpesaCollectionResponse> {
    const username = params.username || this.client.defaults.username;
    const callbackUrl = params.callbackUrl || this.client.defaults.callbackUrl;

    validateRequired({ username, callbackUrl, narration: params.narration }, [
      "username",
      "callbackUrl",
      "narration",
    ]);

    // Normalize +254 to 0-prefix
    const phoneNumber = normalizeKenyaPhone(params.phoneNumber);
    validateKenyaPhone(phoneNumber);
    validateMpesaAmount(params.amount);

    const data = await this.client.request({
      method: "POST",
      path: "/api/v2/payments",
      body: {
        username,
        channel: "MPESA",
        amount: params.amount,
        phone_number: phoneNumber,
        narration: params.narration,
        currency: "KES",
        callback_url: callbackUrl,
      },
    });

    return {
      success: data.success as boolean,
      message: data.message as string,
      status: data.status as string,
      paymentMethod: data.payment_method as string,
      transactionReference: resolveTransactionRef(data),
      trackingId: (data.trackingId as string) || "",
      reference: (data.reference as string) || "",
      result: data.result ?? null,
      _raw: data,
    };
  }

  /**
   * Accept a card payment via hosted checkout.
   *
   * Returns a `checkoutUrl` that you redirect the customer to. The customer
   * enters card details on the hosted page. Final status is delivered via webhook.
   *
   * @example
   * ```typescript
   * const result = await payd.collections.card({
   *   username: "my_payd_user",
   *   amount: 500,
   *   phoneNumber: "0700000000",
   *   narration: "Order #1234",
   *   callbackUrl: "https://example.com/webhook",
   * });
   * // Redirect customer to:
   * console.log(result.checkoutUrl);
   * ```
   */
  async card(params: CardCollectionParams): Promise<CardCollectionResponse> {
    const username = params.username || this.client.defaults.username;
    const callbackUrl = params.callbackUrl || this.client.defaults.callbackUrl;

    validateRequired({ username, callbackUrl, narration: params.narration }, [
      "username",
      "callbackUrl",
      "narration",
    ]);

    const phoneNumber = normalizeKenyaPhone(params.phoneNumber);
    validateKenyaPhone(phoneNumber);
    validateCardAmount(params.amount);

    const data = await this.client.request({
      method: "POST",
      path: "/api/v2/payments",
      body: {
        username,
        channel: "card",
        payment_method: "card",
        amount: params.amount,
        phone_number: phoneNumber,
        narration: params.narration,
        currency: "KES",
        callback_url: callbackUrl,
      },
    });

    // Normalize reference key
    const transactionReference = resolveTransactionRef(data);

    // Card responses must include a checkout URL
    const checkoutUrl = data.checkout_url as string;
    if (!checkoutUrl) {
      throw new PaydAPIError(
        200,
        (data.message as string) || "Card collection response missing checkout_url",
        data,
      );
    }

    return {
      success: data.success as boolean,
      message: data.message as string,
      status: data.status as string,
      paymentMethod: data.payment_method as string,
      checkoutUrl,
      transactionReference,
      trackingId: (data.trackingId as string) || "",
      reference: (data.reference as string) || "",
      result: data.result ?? null,
      _raw: data,
    };
  }

  /**
   * Collect a payment via Pan-African mobile money or bank transfer.
   *
   * Requires calling `payd.networks.discover()` first to get the
   * `networkCode` and `channelId` for the target country.
   *
   * @example
   * ```typescript
   * // Step 1: Discover networks for Nigeria
   * const networks = await payd.networks.discover("receipt", "+234");
   * const network = networks.findMobile("OPay");
   *
   * // Step 2: Collect payment
   * const result = await payd.collections.panAfrican({
   *   username: "my_payd_user",
   *   ...network.toPaymentParams(),
   *   accountName: "phone",
   *   amount: 1800,
   *   phoneNumber: "+2340000000000",
   *   accountNumber: "+2340000000000",
   *   narration: "Order #1234",
   *   currency: "NGN",
   *   callbackUrl: "https://example.com/webhook",
   * });
   * ```
   */
  async panAfrican(params: PanAfricanCollectionParams): Promise<PanAfricanCollectionResponse> {
    const username = params.username || this.client.defaults.username;
    const callbackUrl = params.callbackUrl || this.client.defaults.callbackUrl;

    validateRequired(
      {
        username,
        callbackUrl,
        narration: params.narration,
        accountName: params.accountName,
        accountNumber: params.accountNumber,
        networkCode: params.networkCode,
        channelId: params.channelId,
        currency: params.currency,
        transactionChannel: params.transactionChannel,
      },
      [
        "username",
        "callbackUrl",
        "narration",
        "accountName",
        "accountNumber",
        "networkCode",
        "channelId",
        "currency",
        "transactionChannel",
      ],
    );

    validateInternationalPhone(params.phoneNumber);
    validatePositiveAmount(params.amount);
    validateEnum(params.accountName, ["bank", "phone"], "accountName");
    validateEnum(params.transactionChannel, ["bank", "phone"], "transactionChannel");

    const body: Record<string, unknown> = {
      username,
      account_name: params.accountName,
      amount: params.amount,
      phone_number: params.phoneNumber,
      account_number: params.accountNumber,
      network_code: params.networkCode,
      channel_id: params.channelId,
      narration: params.narration,
      currency: params.currency,
      callback_url: callbackUrl,
      transaction_channel: params.transactionChannel,
    };

    if (params.redirectUrl) {
      body.redirect_url = params.redirectUrl;
    }

    const data = await this.client.request({
      method: "POST",
      path: "/api/v3/payments",
      body,
    });

    // Parse optional bank_account from response
    let bankAccount: BankAccount | undefined;
    const rawBank = data.bank_account as Record<string, string> | undefined;
    if (rawBank) {
      bankAccount = {
        name: rawBank.name || "",
        branchCode: rawBank.branch_code || "",
        accountNumber: rawBank.account_number || "",
        accountName: rawBank.account_name || "",
        accountReference: rawBank.account_reference || "",
      };
    }

    return {
      success: data.success as boolean,
      message: data.message as string,
      status: data.status as string,
      paymentMethod: data.payment_method as string,
      transactionReference: resolveTransactionRef(data),
      bankAccount,
      checkoutUrl: data.checkout_url as string | undefined,
      trackingId: (data.trackingId as string) || "",
      reference: (data.reference as string) || "",
      result: data.result ?? null,
      _raw: data,
    };
  }
}

/**
 * Resolve the transaction reference from whichever key the API returns.
 */
function resolveTransactionRef(data: Record<string, unknown>): string {
  return (
    (data.transaction_reference as string) ||
    (data.correlator_id as string) ||
    (data.payd_transaction_ref as string) ||
    ""
  );
}

File: python/src/payd/resources/collections.py
"""Collection (pay-in) operations."""

from __future__ import annotations

from typing import TYPE_CHECKING

from payd.errors import PaydAPIError
from payd.response import resolve_transaction_reference
from payd.validators import (
    normalize_kenya_phone,
    validate_card_amount,
    validate_enum,
    validate_international_phone,
    validate_kenya_phone,
    validate_mpesa_amount,
    validate_positive_amount,
    validate_required,
)

if TYPE_CHECKING:
    from payd.client import PaydClient


class Collections:
    """Collection (pay-in) operations."""

    def __init__(self, client: PaydClient) -> None:
        self._client = client

    def mpesa(
        self,
        amount: int | float,
        phone_number: str,
        narration: str,
        username: str | None = None,
        callback_url: str | None = None,
    ) -> dict:
        resolved_username = username or self._client.defaults.get("username")
        resolved_callback = callback_url or self._client.defaults.get("callback_url")
        validate_required(
            {
                "username": resolved_username,
                "callback_url": resolved_callback,
                "narration": narration,
            },
            ["username", "callback_url", "narration"],
        )
        assert resolved_username is not None and resolved_callback is not None

        normalized_phone = normalize_kenya_phone(phone_number)
        validate_kenya_phone(normalized_phone)
        validate_mpesa_amount(amount)

        data = self._client.request(
            method="POST",
            path="/api/v2/payments",
            body={
                "username": resolved_username,
                "channel": "MPESA",
                "amount": amount,
                "phone_number": normalized_phone,
                "narration": narration,
                "currency": "KES",
                "callback_url": resolved_callback,
            },
        )

        return {
            "success": bool(data.get("success")),
            "message": str(data.get("message") or ""),
            "status": str(data.get("status") or ""),
            "payment_method": str(data.get("payment_method") or ""),
            "transaction_reference": resolve_transaction_reference(data),
            "tracking_id": str(data.get("trackingId") or ""),
            "reference": str(data.get("reference") or ""),
            "result": data.get("result"),
            "_raw": data,
        }

    def card(
        self,
        amount: int | float,
        phone_number: str,
        narration: str,
        username: str | None = None,
        callback_url: str | None = None,
    ) -> dict:
        resolved_username = username or self._client.defaults.get("username")
        resolved_callback = callback_url or self._client.defaults.get("callback_url")
        validate_required(
            {
                "username": resolved_username,
                "callback_url": resolved_callback,
                "narration": narration,
            },
            ["username", "callback_url", "narration"],
        )
        assert resolved_username is not None and resolved_callback is not None

        normalized_phone = normalize_kenya_phone(phone_number)
        validate_kenya_phone(normalized_phone)
        validate_card_amount(amount)

        data = self._client.request(
            method="POST",
            path="/api/v2/payments",
            body={
                "username": resolved_username,
                "channel": "card",
                "payment_method": "card",
                "amount": amount,
                "phone_number": normalized_phone,
                "narration": narration,
                "currency": "KES",
                "callback_url": resolved_callback,
            },
        )

        checkout_url = data.get("checkout_url")
        if not checkout_url:
            raise PaydAPIError(
                200,
                str(data.get("message") or "Card collection response missing checkout_url"),
                detail=data,
            )

        return {
            "success": bool(data.get("success")),
            "message": str(data.get("message") or ""),
            "status": str(data.get("status") or ""),
            "payment_method": str(data.get("payment_method") or ""),
            "checkout_url": str(checkout_url),
            "transaction_reference": resolve_transaction_reference(data),
            "tracking_id": str(data.get("trackingId") or ""),
            "reference": str(data.get("reference") or ""),
            "result": data.get("result"),
            "_raw": data,
        }

    def pan_african(
        self,
        account_name: str,
        amount: int | float,
        phone_number: str,
        account_number: str,
        network_code: str,
        channel_id: str,
        narration: str,
        currency: str,
        transaction_channel: str,
        username: str | None = None,
        callback_url: str | None = None,
        redirect_url: str | None = None,
        provider_name: str | None = None,
        provider_code: str | None = None,
    ) -> dict:
        resolved_username = username or self._client.defaults.get("username")
        resolved_callback = callback_url or self._client.defaults.get("callback_url")
        validate_required(
            {
                "username": resolved_username,
                "callback_url": resolved_callback,
                "narration": narration,
                "account_name": account_name,
                "account_number": account_number,
                "network_code": network_code,
                "channel_id": channel_id,
                "currency": currency,
                "transaction_channel": transaction_channel,
            },
            [
                "username",
                "callback_url",
                "narration",
                "account_name",
                "account_number",
                "network_code",
                "channel_id",
                "currency",
                "transaction_channel",
            ],
        )
        assert resolved_username is not None and resolved_callback is not None

        validate_international_phone(phone_number)
        validate_positive_amount(amount)
        validate_enum(account_name, ["bank", "phone"], "account_name")
        validate_enum(transaction_channel, ["bank", "phone"], "transaction_channel")

        body: dict[str, str | int | float] = {
            "username": resolved_username,
            "account_name": account_name,
            "amount": amount,
            "phone_number": phone_number,
            "account_number": account_number,
            "network_code": network_code,
            "channel_id": channel_id,
            "narration": narration,
            "currency": currency,
            "callback_url": resolved_callback,
            "transaction_channel": transaction_channel,
        }
        if redirect_url:
            body["redirect_url"] = redirect_url

        data = self._client.request(method="POST", path="/api/v3/payments", body=body)

        bank_account = None
        raw_bank = data.get("bank_account")
        if isinstance(raw_bank, dict):
            bank_account = {
                "name": str(raw_bank.get("name") or ""),
                "branch_code": str(raw_bank.get("branch_code") or ""),
                "account_number": str(raw_bank.get("account_number") or ""),
                "account_name": str(raw_bank.get("account_name") or ""),
                "account_reference": str(raw_bank.get("account_reference") or ""),
            }

        co = data.get("checkout_url")
        return {
            "success": bool(data.get("success")),
            "message": str(data.get("message") or ""),
            "status": str(data.get("status") or ""),
            "payment_method": str(data.get("payment_method") or ""),
            "transaction_reference": resolve_transaction_reference(data),
            "bank_account": bank_account,
            "checkout_url": str(co) if co else None,
            "tracking_id": str(data.get("trackingId") or ""),
            "reference": str(data.get("reference") or ""),
            "result": data.get("result"),
            "_raw": data,
        }
